### PR TITLE
Task 42: FE: Cart: Implement billing info section 

### PR DESCRIPTION
**Pull request by di-kinosian**

### CODE REVIEW

File: src/components/Cart/BillingForm.tsx [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+
+  useEffect(() => {
+    dispatch(getCountries());
+  }, []);
+
+  useEffect(() => {
+    if (country) {
+      dispatch(getCities({ country: country }));
+    } else {
+      setValue("city", "");
```
Comment (VYaniv) [inline src/components/Cart/BillingForm.tsx:53]: I think it is better to use `resetField` here

File: src/components/Cart/BillingForm.tsx [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+  return (
+    <form
+      onSubmit={handleSubmit(onSubmit)}
+      className="flex flex-col gap-7 max-w-[500px]"
+    >
+      <div className="flex flex-col gap-4">
+        <FormField
+          label="First name"
+          placeholder="firstName"
```
Comment (VYaniv) [inline src/components/Cart/BillingForm.tsx:77]: Do you really need placeholders in camel case? It is better to have like a label

File: src/components/Cart/BillingForm.tsx [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+                  placeholder={country ? "city" : "select country"}
+                  disabled={!country}
+                />
+                {errors.city && (
+                  <span className="text-red-500">{errors.city.message}</span>
+                )}
+              </FormControl>
+            );
+          }}
+        />
```
Comment (VYaniv) [inline src/components/Cart/BillingForm.tsx:153]: Put autocomplete into separate reusable component please. And try to use the same approach everywhere, via `Controller` or `register`

File: src/components/FormComponents/Autocomplete.tsx [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+          <TextField {...params} placeholder={placeholder} />
+        )}
+        classes={{ root: classes.root }}
+        onChange={(_, newValue) => {
+          handleChange(newValue?.value as string);
+        }}
+      />
+    </>
+  );
+};
```
Comment (VYaniv) [inline src/components/FormComponents/Autocomplete.tsx:51]: Do you have two autocompletes?

File: src/main/helpers.ts [hunk 5]
```diff
@@ -32,9 +34,25 @@ export const calculateOriginalPrice = (
...
+    label: country.name,
+  }));
+};
+
+export const transformCitiesToOptions = (cities: string[]): Options[] => {
+  return cities.map((city) => ({
+    value: city.toLowerCase(),
+    label: city,
+  }));
+};
```
Comment (VYaniv) [inline src/main/helpers.ts:58]: It also can be merged into one universal function. But it is up to you

File: src/redux/features/location/locationSlice.ts [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+export const getCountries = createAsyncThunk<Country[], void, ThunkRejectValue>(
+  "countries/getCountries",
+  async (_, thunkAPI) => {
+    try {
+      const response = await api.get(
+        "https://countriesnow.space/api/v0.1/countries/capital",
+      );
+
+      return [...response.data.data];
```
Comment (VYaniv) [inline src/redux/features/location/locationSlice.ts:22]: It should call api to your BE and then BE call fake api

File: src/redux/features/location/locationSlice.ts [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+  string[],
+  { country: string },
+  ThunkRejectValue
+>("countries/getCities", async (payload, thunkAPI) => {
+  try {
+    const response = await api.post(
+      "https://countriesnow.space/api/v0.1/countries/cities",
+      {
+        country: payload.country,
+      },
```
Comment (VYaniv) [inline src/redux/features/location/locationSlice.ts:39]: The same as above

File: src/components/Cart/BillingForm.tsx (hunk 1 above)
Comment (di-kinosian) [reply src/components/Cart/BillingForm.tsx:53]: done

File: src/components/FormComponents/Autocomplete.tsx (hunk 4 above)
Comment (di-kinosian) [reply src/components/FormComponents/Autocomplete.tsx:51]: no, I use separate component

File: src/redux/features/location/locationSlice.ts (hunk 6 above)
Comment (di-kinosian) [reply src/redux/features/location/locationSlice.ts:22]: done

File: src/redux/features/location/locationSlice.ts (hunk 7 above)
Comment (di-kinosian) [reply src/redux/features/location/locationSlice.ts:39]: done

File: src/components/Cart/BillingForm.tsx (hunk 2 above)
Comment (di-kinosian) [reply src/components/Cart/BillingForm.tsx:77]: done

File: src/components/Cart/BillingForm.tsx (hunk 3 above)
Comment (di-kinosian) [reply src/components/Cart/BillingForm.tsx:153]: Autocomplete is already separated in the next task. Controller approach is done in this task.

File: src/components/Cart/BillingForm.tsx [hunk 8]
```diff
@@ -66,42 +70,40 @@ export const BillingForm = () => {
...
+          <FormField
+            label={label}
+            placeholder={label}
+            {...field}
+            error={fieldState.error?.message}
+          />
+        )}
+      />
+    );
+  };
```
Comment (VYaniv) [inline src/components/Cart/BillingForm.tsx:88]: Why do you need this wrapper here?

Comment (di-kinosian) [reply src/components/Cart/BillingForm.tsx:88]: @VYaniv , do you mean BillingFormField? It helps reduce repetitive code.

Comment (VYaniv) [reply src/components/Cart/BillingForm.tsx:88]: Yes, but let`s discuss it

**Review by VYaniv — APPROVED** [review]

**Pull request merged**